integrations/spark/spark-3.5/openhouse-spark-itest/src/test/java/com/linkedin/openhouse/spark/catalogtest/BranchTestSpark3_5.java

```diff
@@ -0,0 +1,2370 @@
+
+      // Verify no branches exist yet (WAP staging doesn't create branches)
+      List<Row> refsAfterWapStaging =
+          spark.sql("SELECT name FROM " + tableName + ".refs").collectAsList();
+      assertEquals(0, refsAfterWapStaging.size(), "WAP staging should not create branches");
+
+      // Verify WAP data is not visible in main queries (no branch exists)
+      assertEquals(
+          0,
+          spark.sql("SELECT * FROM " + tableName + "").collectAsList().size(),
+          "Should see 0 rows - no branches exist, WAP data is staged");
+
+      // ===== WAP PUBLISHING TO CREATE MAIN BRANCH =====
+
+      // 2. Publish WAP data to create main branch
+      String wapSnapshotId = String.valueOf(wapSnapshots.get(0).getLong(0));
+      spark.sql(
+          "CALL openhouse.system.cherrypick_snapshot('"
+              + tableName.replace("openhouse.", "")
+              + "', "
+              + wapSnapshotId
+              + ")");
+
+      // Verify main branch now exists
+      List<Row> refsAfterPublishing =
+          spark.sql("SELECT name FROM " + tableName + ".refs ORDER BY name").collectAsList();
+      assertEquals(
+          1, refsAfterPublishing.size(), "Should have main branch after publishing WAP data");
+      assertEquals("main", refsAfterPublishing.get(0).getString(0), "Should have main branch");
+
+      // Verify WAP data is now visible in main branch
+      assertEquals(
+          1,
+          spark.sql("SELECT * FROM " + tableName + "").collectAsList().size(),
+          "Main branch should have 1 row after WAP publishing");
+
+      List<Row> mainData = spark.sql("SELECT name FROM " + tableName + "").collectAsList();
+      assertEquals(
+          "wap_staged_data_1", mainData.get(0).getString(0), "Should see published WAP data");
+
+      // ===== MULTI-WAP OPERATIONS =====
+
+      // 3. Create multiple WAP staged data sets
+      spark.conf().set("spark.wap.id", "wap-stage-2");
+      spark.sql("INSERT INTO " + tableName + " VALUES ('wap_staged_data_2')");
+      spark.conf().unset("spark.wap.id");
+
+      spark.conf().set("spark.wap.id", "wap-stage-3");
+      spark.sql("INSERT INTO " + tableName + " VALUES ('wap_staged_data_3')");
+      spark.conf().unset("spark.wap.id");
+
+      // Verify multiple WAP snapshots exist
+      List<Row> allWapSnapshots =
+          spark
+              .sql(
+                  "SELECT snapshot_id FROM "
+                      + tableName
+                      + ".snapshots "
+                      + "WHERE summary['wap.id'] IS NOT NULL")
+              .collectAsList();
+      assertEquals(3, allWapSnapshots.size(), "Should have 3 WAP staged snapshots");
+
+      // Verify main branch is unchanged (WAP data is staged)
+      assertEquals(
+          1,
+          spark.sql("SELECT * FROM " + tableName + "").collectAsList().size(),
+          "Main branch should still have 1 row (staged WAP not visible)");
+
+      // ===== SELECTIVE WAP PUBLISHING =====
+
```

Comment:
swap the publish order here, wap3 before wap2, then do a time-travel query see if the cherry pick order is maintained.